nexus/src/app/sagas/region_replacement_drive.rs

```diff
@@ -465,17 +465,69 @@ async fn check_from_previous_propolis_step(
         return Ok(DriveCheck::ActionRequired);
     };
 
+    // `migration_id` is set at the beginning of an instance migration (before
+    // anything has happened), and is cleared at the end (after the migration is
+    // finished but before the migration target activates disk Volumes). For
+    // now, return `DriveCheck::Wait`, and pick up driving the region
+    // replacement forward after the migration has completed.
+    //
+    // If this saga does not wait, it will interleave with the instance
+    // migration saga. Depending on Nexus' view of what stage the migration is
+    // in, volume replacement requests could be sent to the source propolis or
+    // destination propolis.
+    //
+    // Processing a replacement request does _not_ cause an activation, so
+    // sending a replacement request to the source propolis will not cause the
+    // destination to be unable to activate (even though the destination _could_
+    // be using a VCR with a lower generation number than what the replacement
+    // request has!). It will probably cause live repair to start on the source,
+    // which is alright because it can be cancelled at any time (and will be
+    // when the destination propolis activates the Volume).
+    //
+    // Until crucible#871 is addressed, sending the replacement request to the
+    // destination propolis could cause a panic if activation hasn't occurred
+    // yet. Even if this saga does wait, this same potential exists because the
+    // migration is considered complete before propolis activates disk Volumes.
+    //
+    // If the destination propolis' Volume activated, the Upstairs will return a
+    // `ReplacementResult`: either `VcrMatches` (if the destination is using the
+    // updated VCR) or `Started` (if the destination is using the pre-update VCR
+    // and the replacement result triggers live repair).
+    //
+    // Also note: if the migration target was sent a Volume that refers to a
+    // region that is no longer responding, it will hang trying to activate, but
+    // the migration itself will succeed (clearing the migration ID!). This is
+    // especially bad because it's easy to hit: if a region goes away and a
+    // migration is triggered before the region replacement start saga can swap
+    // out the region that's gone, the migration saga will checkout the
+    // pre-update Volume and the destination propolis will hit this scenario.
```

Comment:
As we discussed at today's huddle, I think this implies that we need (1) a mechanism for the Upstairs to accept a region replacement request while it's trying to connect during activation, and (2) assurance that Propolis will actually forward such a request to a Crucible upstairs if one arrives (i.e. if an upstairs is in the "spinning waiting to activate" state Propolis still needs to be able to forward it a configuration change request). I'll have a glance at (2) as part of the Propolis work I'm currently doing.

If there's a Crucible issue for (1) I would be inclined to link it here.

Reply:
Crucible issue for replace before active: https://github.com/oxidecomputer/crucible/issues/871

Reply:
^